src/features/registry.js

```diff
@@ -21,43 +21,67 @@ systemJSPrototype.set = function (id, module) {
     if (toStringTag)
       Object.defineProperty(ns, toStringTag, { value: 'Module' });
   }
+
   const done = Promise.resolve(ns);
-  this.delete(id);
-  this[REGISTRY][id] = {
+
+  const load = this[REGISTRY][id] || (this[REGISTRY][id] = {
     id: id,
     i: [],
-    n: ns,
-    I: done,
-    L: done,
     h: false,
     d: [],
     e: null,
     er: undefined,
-    E: undefined,
+    E: undefined
+  });
+
+  if (load.e || load.E)
+    return false;
+  
+  Object.assign(load, {
+    n: ns,
+    I: done,
+    L: done,
     C: done
-  };
+  });
   return ns;
 };
 
 systemJSPrototype.has = function (id) {
   const load = this[REGISTRY][id];
-  return load && load.e === null && !load.E;
+  return !!load;
 };
 
 // Delete function provided for hot-reloading use cases
 systemJSPrototype.delete = function (id) {
-  const load = this.get(id);
-  if (load === undefined)
+  const registry = this[REGISTRY];
+  const load = registry[id];
+  // in future we can support load.E case by failing load first
+  // but that will require TLA callbacks to be implemented
+  if (!load || load.e !== null || load.E)
     return false;
+
+  const importerSetters = load.i;
   // remove from importerSetters
   // (release for gc)
-  if (load && load.d)
+  if (load.d)
     load.d.forEach(function (depLoad) {
       const importerIndex = depLoad.i.indexOf(load);
       if (importerIndex !== -1)
         depLoad.i.splice(importerIndex, 1);
     });
-  return delete this[REGISTRY][id];
+  delete registry[id];
+  return function () {
+    const load = registry[id];
+    if (!load || load.e !== null || load.E)
+      return false;
+    // add back the old setters
+    load.i.forEach(setter => {
```

Comment:
Should this not be:
```js
    importerSetters.forEach(setter => {
      if (load.i.indexOf(setter) === -1) {
        load.i.push(setter);
        setter(load.n);
      }
    });
```

is the indexOf check even needed, can the same setter be set initially>

Reply:
Yes, you are right on both.

I think the only way the setters would already be there is if you ran the update function twice, so I've added a simple guard to that.